libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs

```diff
@@ -85,88 +91,106 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
+                        : opt.InlineSeparator ?? string.Empty;
+                    index++;
                 }
             }
-            txt += "";
 
             // Return activity with choices as an inline list.
-            return MessageFactory.Text(txt, speak, InputHints.ExpectingInput);
+            return MessageFactory.Text(sb.ToString(), speak, InputHints.ExpectingInput);
         }
 
-        public static Activity List(List<string> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
+        /// <summary>
+        /// Gets a "normalized" title for a <see cref="Choice"/> object.
+        /// </summary>
+        /// <param name="choice">The choice object.</param>
+        /// <returns>The normalized title.</returns>
+        /// <exception cref="InvalidOperationException">The choice does not have a valid action title or value to use as a title.</exception>
+        private static string GetTitle(Choice choice)
+        {
+            if (choice is null) return null;
+
+            if (!string.IsNullOrWhiteSpace(choice.Action?.Title)) return choice.Action.Title.Trim();
+            if (!string.IsNullOrWhiteSpace(choice.Value)) return choice.Value.Trim();
+
+            throw new InvalidOperationException("Each choice must specify either an action title or a value.");
+        }
+
+        public static Activity List(IEnumerable<string> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
         {
             return List(ToChoices(choices), text, speak, options);
         }
 
-        public static Activity List(List<Choice> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
+        public static Activity List(IEnumerable<Choice> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
         {
-            choices = choices ?? new List<Choice>();
             options = options ?? new ChoiceFactoryOptions();
 
             bool includeNumbers = options.IncludeNumbers ?? true;
 
             // Format list of choices
-            var connector = string.Empty;
-            var txt = (text ?? string.Empty);
-            txt += "\n\n   ";
-
-            for (var index = 0; index < choices.Count; index++)
+            var separator = string.Empty;
+            var sb = new StringBuilder();
+            if (text != null)
             {
-                var choice = choices[index];
-
-                var title = choice.Action != null && choice.Action.Title != null ? choice.Action.Title : choice.Value;
+                sb.Append(text);
+            }
+            sb.Append(" ");
 
-                txt += connector;
-                if (includeNumbers)
-                {
-                    txt += (index + 1).ToString() + ". ";
-                }
-                else
+            if (choices != null)
+            {
+                int count = choices.Where(c => c != null).Count();
```

Comment:
See comment above about `.Count()` and double enumeration, but in this case there's truly no need to do the `.Count()` because `count` doesn't even appear to be used in this logic, right???